src/MarginTrading.MarketMaker/Infrastructure/Implementation/TraceService.cs

```diff
@@ -0,0 +1,130 @@
+﻿using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using Lykke.Logs;
+using MarginTrading.MarketMaker.Contracts.Models;
+using MarginTrading.MarketMaker.Enums;
+using Newtonsoft.Json;
+using Newtonsoft.Json.Converters;
+
+namespace MarginTrading.MarketMaker.Infrastructure.Implementation
+{
+    internal class TraceService : ITraceService, ICustomStartup
+    {
+        private static readonly JsonSerializerSettings JsonSerializerSettings = new JsonSerializerSettings
+        {
+            Converters = {new StringEnumConverter()}
+        };
+
+        private static readonly BlockingCollection<TraceMessage> WritingQueue =
+            new BlockingCollection<TraceMessage>(10000);
+
+        private static readonly
+            ConcurrentDictionary<(TraceLevelGroupEnum Group, string AssetPairId), ConcurrentQueue<TraceMessage>>
+            LastElemsQueues
+                = new ConcurrentDictionary<(TraceLevelGroupEnum, string), ConcurrentQueue<TraceMessage>>();
+
+        private readonly ISystem _system;
+        private readonly LykkeLogToAzureStorage _logToAzureStorage;
+
+        public TraceService(ISystem system,
+            LykkeLogToAzureStorage logToAzureStorage)
+        {
+            _system = system;
+            _logToAzureStorage = logToAzureStorage;
+        }
+
+        public void Initialize()
+        {
+            Task.Run(() =>
+            {
+                while (true)
+                    try
+                    {
+                        foreach (var m in WritingQueue.GetConsumingEnumerable())
+                        {
+                            var lastElemsQueue =
+                                LastElemsQueues.GetOrAdd((m.TraceGroup, m.AssetPairId),
+                                    k => new ConcurrentQueue<TraceMessage>());
+                            lastElemsQueue.Enqueue(m);
+
+                            while (lastElemsQueue.Count > 100)
+                                lastElemsQueue.TryDequeue(out var _);
+
+                            var message = m.AssetPairId + '\t' + m.TraceGroup + '\t' + m.Msg;
+                            Console.WriteLine(message);
+                            _logToAzureStorage.WriteInfoAsync("MtMmTrace",
+                                JsonConvert.SerializeObject(m, JsonSerializerSettings), message);
+                        }
+                    }
+                    catch (Exception e)
+                    {
+                        Console.WriteLine(e);
```

Comment:
And don't need this error in any log?

Reply:
For example?

Reply:
I added slack logs here, but in case we cannot reach azure this won't help